Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs

```diff
@@ -0,0 +1,321 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE."
+**/
+
+using Microsoft.Alm.Authentication;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Trace = Microsoft.Alm.Git.Trace;
+
+namespace Atlassian.Bitbucket.Authentication.OAuth
+{
+
+    /// <summary>
+    /// </summary>
+    public class OAuthAuthenticator
+    {
+        /// <summary>
+        /// The maximum wait time for a network request before timing out
+        /// </summary>
+        public const int RequestTimeout = 15 * 1000; // 15 second limit
+
+        public OAuthAuthenticator()
+        {
+        }
+
+        public string AuthorizeUri { get { return "/site/oauth2/authorize"; } }
+        public string TokenUri { get { return "/site/oauth2/access_token"; } }
+
+        public string ConsumerKey { get { return "HJdmKXV87DsmC9zSWB"; } }
+        public string ConsumerSecret { get { return "wwWw47VB9ZHwMsD4Q4rAveHkbxNrMp3n"; } }
+        public string CallbackUri { get { return "http://localhost:34106/"; } }
+
+        /// <summary>
+        ///     Gets the OAuth access token
+        /// </summary>
+        /// <returns>The access token</returns>
+        /// <exception cref="SourceTree.Exceptions.OAuthException">Thrown when OAuth fails for whatever reason</exception>
+        public async Task<AuthenticationResult> GetAuthAsync(TargetUri targetUri, TokenScope scope, CancellationToken cancellationToken)
+        {
+
+            var authToken = await Authorize(targetUri, scope, cancellationToken);
+
+            return await GetAccessToken(targetUri, authToken);
+        }
+
+        /// <summary>
+        ///     Uses a refresh_token to get a new access_token
+        /// </summary>
+        /// <param name="targetUri"></param>
+        /// <param name="refreshToken"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<AuthenticationResult> RefreshAuthAsync(TargetUri targetUri, string refreshToken, CancellationToken cancellationToken)
+        {
+            return await RefreshAccessToken(targetUri, refreshToken);
+        }
+
+        /// <summary>
+        ///     Run the OAuth dance to get a new request_token
+        /// </summary>
+        /// <param name="targetUri"></param>
+        /// <param name="scope"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task<string> Authorize(TargetUri targetUri, TokenScope scope, CancellationToken cancellationToken)
+        {
+            var authorityUrl = string.Format(
```

Comment:
nit: this method call is rather difficult to read, any chance you can clean it (and the other below) up?